VoodooPS2Trackpad/VoodooPS2Elan.cpp

```diff
@@ -0,0 +1,2219 @@
+#else
+#define INTERRUPT_LOG(args...)  do { } while (0)
+#endif
+
+#include "LegacyIOService.h"
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Winconsistent-missing-override"
+#include <IOKit/IOLib.h>
+#include <IOKit/hidsystem/IOHIDParameter.h>
+#include <IOKit/IOWorkLoop.h>
+#include <IOKit/IOTimerEventSource.h>
+#include <IOKit/usb/IOUSBHostFamily.h>
+#include <IOKit/usb/IOUSBHostHIDDevice.h>
+#include <IOKit/bluetooth/BluetoothAssignedNumbers.h>
+#pragma clang diagnostic pop
+#include "VoodooPS2Controller.h"
+#include "VoodooPS2Elan.h"
+#include "VoodooInputMultitouch/VoodooInputTransducer.h"
+#include "VoodooInputMultitouch/VoodooInputMessages.h"
+
+// =============================================================================
+// ApplePS2Elan Class Implementation
+//
+
+OSDefineMetaClassAndStructors(ApplePS2Elan, IOHIPointing);
+
+UInt32 ApplePS2Elan::deviceType()
+{ return NX_EVS_DEVICE_TYPE_MOUSE; };
+
+UInt32 ApplePS2Elan::interfaceID()
+{ return NX_EVS_DEVICE_INTERFACE_BUS_ACE; };
+
+#define abs(x) ((x) < 0 ? -(x) : (x))
+
+bool ApplePS2Elan::init(OSDictionary *dict) {
+    // Initialize this object's minimal state. This is invoked right after this
+    // object is instantiated.
+
+    if (!super::init(dict)) {
+        return false;
+    }
+
+    // announce version
+    extern kmod_info_t kmod_info;
+    DEBUG_LOG("VoodooPS2Elan: Version %s starting on OS X Darwin %d.%d.\n", kmod_info.version, version_major, version_minor);
+
+    return true;
+}
+
+void ApplePS2Elan::injectVersionDependentProperties(OSDictionary *config) {
+    // inject properties specific to the version of Darwin that is runnning...
+    char buf[32];
+    OSDictionary *dict = NULL;
+    do {
+        // check for "Darwin major.minor"
+        snprintf(buf, sizeof(buf), "Darwin %d.%d", version_major, version_minor);
+        if ((dict = OSDynamicCast(OSDictionary, config->getObject(buf)))) {
+            break;
+        }
+
+        // check for "Darwin major.x"
+        snprintf(buf, sizeof(buf), "Darwin %d.x", version_major);
+        if ((dict = OSDynamicCast(OSDictionary, config->getObject(buf)))) {
+            break;
+        }
+
+        // check for "Darwin 16+" (this is what is used currently, other formats are for future)
+        if (version_major >= 16 && (dict = OSDynamicCast(OSDictionary, config->getObject("Darwin 16+")))) {
+            break;
+        }
+    } while (0);
+
+    if (dict) {
+        // found version specific properties above, inject...
+        if (OSCollectionIterator *iter = OSCollectionIterator::withCollection(dict)) {
+            // Note: OSDictionary always contains OSSymbol*
+            while (const OSSymbol *key = static_cast<const OSSymbol*>(iter->getNextObject())) {
+                if (OSObject *value = dict->getObject(key)) {
+                    setProperty(key, value);
+                }
+            }
+            iter->release();
+        }
+    }
+}
```

Comment:
This is a rudiment left from Rehabman's work. He decided not to enable the new prefpane on older systems (don't know why) so added this version check. As we are simulating MT2, we don't need to support systems that don't support MT2, and need to add properties on all supported systems.

But as it's not removed for Synaptics, we can merge it as is and then remove this code in both classes.